<a href="https://colab.research.google.com/github/WilliamAshbee/DatascienceHomeworks/blob/main/HW1_Ashbee_William.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Homework 1 - Data Exploration using Census Data

In this homework assignment you will use census data from the 1994 and 1995 current population surveys conducted by the U.S. Census Bureau. A copy of this dataset is located at UCI Machine Learning repository, please see this [link](https://archive.ics.uci.edu/ml/datasets/Census-Income+%28KDD%29) to reach to the dataset website.

This dataset contains census data extracted from the 1994 and 1995 Current Population Surveys. We will only work with 'training' data (a link to that is provided to you below). The data contains 41 demographic and employment related variables. The abbreviated column names provided to you below. You are expected to read the documentation of this dataset, understand the features and preprocess this dataset. Additional information can be found [in the data description](https://archive.ics.uci.edu/ml/machine-learning-databases/census-income-mld/census-income.data.html) and [additional comments](https://archive.ics.uci.edu/ml/machine-learning-databases/census-income-mld/census-income.names).

Below you will find a code snippet to download and read the data into a pandas dataframe. You can alternatively download it yourself, extract and read it manually. The questions are shown in the subsequent cells. You can provide your answers in this file. Make sure to change the notebook's name and add your name before submitting.

In [1]:
import pandas as pd
import numpy as np

In [2]:
import urllib.request
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/census-income-mld/census-income.data.gz'
census_dl_file = './census-income.data.gz'
urllib.request.urlretrieve(url, census_dl_file)

('./census-income.data.gz', <http.client.HTTPMessage at 0x7f62fadd7950>)

In [3]:
columns_abbr = ['AAGE', 'ACLSWKR', 'ADTIND', 'ADTOCC', 'AHGA', 'AHRSPAY', 'AHSCOL', 'AMARITL', 'AMJIND', 'AMJOCC',
                'ARACE', 'AREORGN', 'ASEX', 'AUNMEM', 'AUNTYPE', 'AWKSTAT', 'CAPGAIN', 'CAPLOSS', 'DIVVAL', 
                'FILESTAT', 'GRINREG', 'GRINST', 'HHDFMX', 'HHDREL', 'MARSUPWT', 'MIGMTR1', 'MIGMTR3', 'MIGMTR4', 
                'MIGSAME', 'MIGSUN', 'NOEMP', 'PARENT', 'PEFNTVTY', 'PEMNTVTY', 'PENATVTY', 'PRCITSHP', 'SEOTR', 
                'VETQVA', 'VETYN', 'WKSWORK', 'YEAR', 'PTOTVALB']

In [4]:
df = pd.read_csv('census-income.data.gz', compression='gzip', names=columns_abbr, sep=r',', skipinitialspace=True)
df.head()

AAGE                         ACLSWKR  ADTIND  ...  WKSWORK YEAR  PTOTVALB
0    73                 Not in universe       0  ...        0   95  - 50000.
1    58  Self-employed-not incorporated       4  ...       52   94  - 50000.
2    18                 Not in universe       0  ...        0   95  - 50000.
3     9                 Not in universe       0  ...        0   94  - 50000.
4    10                 Not in universe       0  ...        0   94  - 50000.

[5 rows x 42 columns]

In [5]:
# you can see the DataFrame's info panel here
columns_names = ['AAGE',
                 'ACLSWKR',
                 'ADTIND',
                 'ADTOCC',
                 'AHGA',
                 'AHRSPAY',
                 'AHSCOL', 
                 'AMARITL',
                 'AMJIND', 
                 'AMJOCC',
                'ARACE', 
                 'AREORGN',
                 'ASEX',
                 'AUNMEM', 
                 'AUNTYPE',
                 'AWKSTAT',
                 'CAPGAIN', 
                 'CAPLOSS', 
                 'DIVVAL',
                'FILESTAT',
                 'GRINREG',
                 'GRINST', 
                 'HHDFMX', 
                 'HHDREL', 
                 'MARSUPWT', 
                 'MIGMTR1',
                 'MIGMTR3', 
                 'MIGMTR4',
                'MIGSAME', 
                 'MIGSUN',
                 'NOEMP', 
                 'PARENT', 
                 'PEFNTVTY', 
                 'PEMNTVTY', 
                 'PENATVTY', 
                 'PRCITSHP', 
                 'SEOTR', 
                'VETQVA', 
                 'VETYN', 
                 'WKSWORK', 
                 'YEAR', 
                 'PTOTVALB']

column_desc = {'AAGE':'Age',
'ACLSWKR':'class of worker',
'ADTIND':'industry code',
'ADTOCC':'occupation code',
'AHGA':'education',
'AHRSPAY':'wage per hour',
'AHSCOL':'enrolled in edu inst last wk',
'AMARITL':'marital status',
'AMJIND':'major industry code',
'AMJOCC':'major occupation code',
'ARACE':'mace',
'AREORGN':'hispanic Origin',
'ASEX':'sex',
'AUNMEM':'member of a labor union',
'AUNTYPE':'reason for unemployment',
'AWKSTAT':'full or part time employment stat',
'CAPGAIN':'capital gains',
'CAPLOSS':'capital losses',
'DIVVAL':'divdends from stocks',
'FILESTAT':'tax filer status',
'GRINREG':'region of previous residence',
'GRINST':'state of previous residence',
'HHDFMX':'detailed household and family stat',
'HHDREL':'detailed household summary in household',
'MARSUPWT':'instance weight',
'MIGMTR1':'migration code-change in msa',
'MIGMTR3':'migration code-change in reg',
'MIGMTR4':'migration code-move within reg',
'MIGSAME':'live in this house 1 year ago',
'MIGSUN':'migration prev res in sunbelt',
'NOEMP':'num persons worked for employer',
'PARENT':'family members under 18',
'PEFNTVTY':'country of birth father',
'PEMNTVTY':'country of birth mother',
'PENATVTY':'country of birth self',
'PRCITSHP':'citizenship',
'SEOTR':'own business or self employed',
'VETQVA':'fill inc questionnaire for veterans admin',
'VETYN':'veterans benefits',
'WKSWORK':'weeks worked in year',
'YEAR':'YEAR',
'PTOTVALB':'PTOTVALB'}

a = set(columns_names)
b = set(column_desc.keys())
print(a-b)
print(len(columns_names),len(column_desc))
df.info()

set()
42 42
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 199523 entries, 0 to 199522
Data columns (total 42 columns):
 #   Column    Non-Null Count   Dtype  
---  ------    --------------   -----  
 0   AAGE      199523 non-null  int64  
 1   ACLSWKR   199523 non-null  object 
 2   ADTIND    199523 non-null  int64  
 3   ADTOCC    199523 non-null  int64  
 4   AHGA      199523 non-null  object 
 5   AHRSPAY   199523 non-null  int64  
 6   AHSCOL    199523 non-null  object 
 7   AMARITL   199523 non-null  object 
 8   AMJIND    199523 non-null  object 
 9   AMJOCC    199523 non-null  object 
 10  ARACE     199523 non-null  object 
 11  AREORGN   198649 non-null  object 
 12  ASEX      199523 non-null  object 
 13  AUNMEM    199523 non-null  object 
 14  AUNTYPE   199523 non-null  object 
 15  AWKSTAT   199523 non-null  object 
 16  CAPGAIN   199523 non-null  int64  
 17  CAPLOSS   199523 non-null  int64  
 18  DIVVAL    199523 non-null  int64  
 19  FILESTAT  199523 non-null  objec

### Q1 - Identify the data scales and data types for each variable in census data. Identify the domain for each variable by checking the attributes' values. Then, create a data quality report for both categorical (nominal, ordinal) and continuous (interval, ratio) variables. [40 pts]

For data scales, identify whether an attribute is nominal, ordinal, interval or ratio scale.
For data types, identify the domain and provide an appropriate data type (integer, float, String, date, Boolean). See if that data type is correct in your dataframe.
For domain, inspect each distinct value for each attribute. Identify missing values. Also include the bar plots and histograms for visualizing the distributions. 

*The examples for a continious and a categorical feature can be seen below. You do not need to use jupyter formatting provided here. You can print a DataFrame or read a csv, and display it. Make sure you have the csv in your zip file.*

In [6]:
# your code goes here
categorical = ['ACLSWKR', 'AHGA', 'AHSCOL', 'AMARITL', 'AMJIND', 'AMJOCC', 'ARACE', 'AREORGN', 'ASEX', 'AUNMEM', 'AUNTYPE', 'AWKSTAT', 'FILESTAT', 'GRINREG', 'GRINST', 'HHDFMX', 'HHDREL', 
'MIGMTR1', 'MIGMTR3', 'MIGMTR4', 'MIGSAME', 'MIGSUN', 'PARENT', 'PEFNTVTY', 'PEMNTVTY', 'PENATVTY', 'PRCITSHP', 'VETQVA', 'PTOTVALB']




In [18]:
#df.sum()
# We will use df.iteritems() as it allows us to iterate through columns
for (name, series) in df.iteritems():
  #print('ANALYZING THE COLUMN:', name)
  #print('\tTotal number of records', series.size)
  #print('\tNumber of missing values', series.isnull().sum())
  #print('\tPercentage of missing values {0}%'.format(((series.isnull().sum()/series.size)*100)) )
  #print('\tNumber of unique values', series.unique().size)
  #print('| Feature | Desc. | Count | % of Missing | Card. | Mode | Mode Freq. | Mode % | 2nd Mode | 2nd Mode Freq. | 2nd Mode Perc | Notes |')
  
  #catagorical = True
  if name in categorical:

    modes= series.value_counts().keys()
    card = series.size
    mode1= modes[0]
    mode2= modes[1]
    mode1count= series.value_counts()[mode1]
    mode2count= series.value_counts()[mode2]
    mode1percent = "{:.2f}".format(mode1count*100.0/float(card))
    mode2percent = "{:.2f}".format(mode2count*100.0/float(card))
  
    print("|",name,"|",
          column_desc[name],"|",
          series.size,"|",
          series.isnull().sum(),"|",
          series.unique().size,"|",
          mode1,"|",mode1count,"|",mode1percent,"|",
          mode2,"|",mode2count,"|",mode2percent
          )
    
  #print(mode1,mode1count,mode1percent,mode2,mode2count,mode2percent)


| ACLSWKR | class of worker | 199523 | 0 | 9 | Not in universe | 100245 | 50.24 | Private | 72028 | 36.10
| AHGA | education | 199523 | 0 | 17 | High school graduate | 48407 | 24.26 | Children | 47422 | 23.77
| AHSCOL | enrolled in edu inst last wk | 199523 | 0 | 3 | Not in universe | 186943 | 93.69 | High school | 6892 | 3.45
| AMARITL | marital status | 199523 | 0 | 7 | Never married | 86485 | 43.35 | Married-civilian spouse present | 84222 | 42.21
| AMJIND | major industry code | 199523 | 0 | 24 | Not in universe or children | 100684 | 50.46 | Retail trade | 17070 | 8.56
| AMJOCC | major occupation code | 199523 | 0 | 15 | Not in universe | 100684 | 50.46 | Adm support including clerical | 14837 | 7.44
| ARACE | mace | 199523 | 0 | 5 | White | 167365 | 83.88 | Black | 20415 | 10.23
| AREORGN | hispanic Origin | 199523 | 874 | 10 | All other | 171907 | 86.16 | Mexican-American | 8079 | 4.05
| ASEX | sex | 199523 | 0 | 2 | Female | 103984 | 52.12 | Male | 95539 | 47.88
| AUNMEM | memb

In [8]:
for (name, series) in df.iteritems():
  
  catagorical = True
  if catagorical:

    
    try:
      min = '{}'.format(series.min())
      q1 = '{}'.format(series.quantile(0.25))
      median = '{}'.format(series.median())
      q3 = '{}'.format(series.quantile(0.75))
      max = '{}'.format(series.max())
      mean = '{:.2f}'.format(series.mean())
      stdev = '{:.2f}'.format(series.std()) 

      #| Feature | Desc. | Count | % of Missing | Card. | Min. | Q1 | Median | Q3 | Max. | Mean | Std. Dev. | Notes |
      
      print("|", name,"|",
            column_desc[name],"|",
            series.size,"|",
            series.isnull().sum(),"|",
            series.unique().size,"|",
            min,"|",
            q1,"|",
            median, "|",
            q3,"|",
            max,"|",
            mean,"|",
            stdev,"|"
            )
    except:
        print(series.name, ' is not continuous')

| AAGE | Age | 199523 | 0 | 91 | 0 | 15.0 | 33.0 | 50.0 | 90 | 34.49 | 22.31 |
ACLSWKR  is not continuous
| ADTIND | industry code | 199523 | 0 | 52 | 0 | 0.0 | 0.0 | 33.0 | 51 | 15.35 | 18.07 |
| ADTOCC | occupation code | 199523 | 0 | 47 | 0 | 0.0 | 0.0 | 26.0 | 46 | 11.31 | 14.45 |
AHGA  is not continuous
| AHRSPAY | wage per hour | 199523 | 0 | 1240 | 0 | 0.0 | 0.0 | 0.0 | 9999 | 55.43 | 274.90 |
AHSCOL  is not continuous
AMARITL  is not continuous
AMJIND  is not continuous
AMJOCC  is not continuous
ARACE  is not continuous
AREORGN  is not continuous
ASEX  is not continuous
AUNMEM  is not continuous
AUNTYPE  is not continuous
AWKSTAT  is not continuous
| CAPGAIN | capital gains | 199523 | 0 | 132 | 0 | 0.0 | 0.0 | 0.0 | 99999 | 434.72 | 4697.53 |
| CAPLOSS | capital losses | 199523 | 0 | 113 | 0 | 0.0 | 0.0 | 0.0 | 4608 | 37.31 | 271.90 |
| DIVVAL | divdends from stocks | 199523 | 0 | 1478 | 0 | 0.0 | 0.0 | 0.0 | 99999 | 197.53 | 1984.16 |
FILESTAT  is not continuous
GRINREG  is no

In [15]:
#the following are not continuous
categorical = ['ACLSWKR', 'AHGA', 'AHSCOL', 'AMARITL', 'AMJIND', 'AMJOCC', 'ARACE', 'AREORGN', 'ASEX', 'AUNMEM', 'AUNTYPE', 'AWKSTAT', 'FILESTAT', 'GRINREG', 'GRINST', 'HHDFMX', 'HHDREL', 
'MIGMTR1', 'MIGMTR3', 'MIGMTR4', 'MIGSAME', 'MIGSUN', 'PARENT', 'PEFNTVTY', 'PEMNTVTY', 'PENATVTY', 'PRCITSHP', 'VETQVA', 'PTOTVALB']


| Feature | Desc. | Count | % of Missing | Card. | Min. | Q1 | Median | Q3 | Max. | Mean | Std. Dev. | Notes |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| AAGE | Age | 199,523 | 0 | 91 | 0 | 15 | 33 | 50 | 90 | 34.49 | 22.31 |  |
| AAGE | Age | 199523 | 0 | 91 | 0 | 15.0 | 33.0 | 50.0 | 90 | 34.49 | 22.31 |
| ADTIND | industry code | 199523 | 0 | 52 | 0 | 0.0 | 0.0 | 33.0 | 51 | 15.35 | 18.07 |
| ADTOCC | occupation code | 199523 | 0 | 47 | 0 | 0.0 | 0.0 | 26.0 | 46 | 11.31 | 14.45 |
| AHRSPAY | wage per hour | 199523 | 0 | 1240 | 0 | 0.0 | 0.0 | 0.0 | 9999 | 55.43 | 274.90 |
| CAPGAIN | capital gains | 199523 | 0 | 132 | 0 | 0.0 | 0.0 | 0.0 | 99999 | 434.72 | 4697.53 |
| CAPLOSS | capital losses | 199523 | 0 | 113 | 0 | 0.0 | 0.0 | 0.0 | 4608 | 37.31 | 271.90 |
| DIVVAL | divdends from stocks | 199523 | 0 | 1478 | 0 | 0.0 | 0.0 | 0.0 | 99999 | 197.53 | 1984.16 |
| MARSUPWT | instance weight | 199523 | 0 | 99800 | 37.87 | 1061.615 | 1618.31 | 2188.61 | 18656.3 | 1740.38 | 993.77 |
| NOEMP | num persons worked for employer | 199523 | 0 | 7 | 0 | 0.0 | 1.0 | 4.0 | 6 | 1.96 | 2.37 |
| SEOTR | own business or self employed | 199523 | 0 | 3 | 0 | 0.0 | 0.0 | 0.0 | 2 | 0.18 | 0.55 |
| VETYN | veterans benefits | 199523 | 0 | 3 | 0 | 2.0 | 2.0 | 2.0 | 2 | 1.51 | 0.85 |
| WKSWORK | weeks worked in year | 199523 | 0 | 53 | 0 | 0.0 | 8.0 | 52.0 | 52 | 23.17 | 24.41 |
| YEAR | YEAR | 199523 | 0 | 2 | 94 | 94.0 | 94.0 | 95.0 | 95 | 94.50 | 0.50 |

1.   List item

*   List item
*   List item

---




2.   List item

| Feature | Desc. | Count | % of Missing | Card. | Mode | Mode Freq. | Mode % | 2nd Mode | 2nd Mode Freq. | 2nd Mode Perc | Notes |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| ACLSWKR | Class of worker | 199,523 | 0 | 9 | Not in Universe | 100,245 | 50.24 | Private | 72,028 |  36.10 |  | *italicized text*
| ACLSWKR | class of worker | 199523 | 0 | 9 | Not in universe | 100245 | 50.24 | Private | 72028 | 36.10
| AHGA | education | 199523 | 0 | 17 | High school graduate | 48407 | 24.26 | Children | 47422 | 23.77
| AHSCOL | enrolled in edu inst last wk | 199523 | 0 | 3 | Not in universe | 186943 | 93.69 | High school | 6892 | 3.45
| AMARITL | marital status | 199523 | 0 | 7 | Never married | 86485 | 43.35 | Married-civilian spouse present | 84222 | 42.21
| AMJIND | major industry code | 199523 | 0 | 24 | Not in universe or children | 100684 | 50.46 | Retail trade | 17070 | 8.56
| AMJOCC | major occupation code | 199523 | 0 | 15 | Not in universe | 100684 | 50.46 | Adm support including clerical | 14837 | 7.44
| ARACE | mace | 199523 | 0 | 5 | White | 167365 | 83.88 | Black | 20415 | 10.23
| AREORGN | hispanic Origin | 199523 | 874 | 10 | All other | 171907 | 86.16 | Mexican-American | 8079 | 4.05
| ASEX | sex | 199523 | 0 | 2 | Female | 103984 | 52.12 | Male | 95539 | 47.88
| AUNMEM | member of a labor union | 199523 | 0 | 3 | Not in universe | 180459 | 90.45 | No | 16034 | 8.04
| AUNTYPE | reason for unemployment | 199523 | 0 | 6 | Not in universe | 193453 | 96.96 | Other job loser | 2038 | 1.02
| AWKSTAT | full or part time employment stat | 199523 | 0 | 8 | Children or Armed Forces | 123769 | 62.03 | Full-time schedules | 40736 | 20.42
| FILESTAT | tax filer status | 199523 | 0 | 6 | Nonfiler | 75094 | 37.64 | Joint both under 65 | 67383 | 33.77
| GRINREG | region of previous residence | 199523 | 0 | 6 | Not in universe | 183750 | 92.09 | South | 4889 | 2.45
| GRINST | state of previous residence | 199523 | 0 | 51 | Not in universe | 183750 | 92.09 | California | 1714 | 0.86
| HHDFMX | detailed household and family stat | 199523 | 0 | 38 | Householder | 53248 | 26.69 | Child <18 never marr not in subfamily | 50326 | 25.22
| HHDREL | detailed household summary in household | 199523 | 0 | 8 | Householder | 75475 | 37.83 | Child under 18 never married | 50426 | 25.27
| MIGMTR1 | migration code-change in msa | 199523 | 0 | 10 | ? | 99696 | 49.97 | Nonmover | 82538 | 41.37
| MIGMTR3 | migration code-change in reg | 199523 | 0 | 9 | ? | 99696 | 49.97 | Nonmover | 82538 | 41.37
| MIGMTR4 | migration code-move within reg | 199523 | 0 | 10 | ? | 99696 | 49.97 | Nonmover | 82538 | 41.37
| MIGSAME | live in this house 1 year ago | 199523 | 0 | 3 | Not in universe under 1 year old | 101212 | 50.73 | Yes | 82538 | 41.37
| MIGSUN | migration prev res in sunbelt | 199523 | 0 | 4 | ? | 99696 | 49.97 | Not in universe | 84054 | 42.13
| PARENT | family members under 18 | 199523 | 0 | 5 | Not in universe | 144232 | 72.29 | Both parents present | 38983 | 19.54
| PEFNTVTY | country of birth father | 199523 | 0 | 43 | United-States | 159163 | 79.77 | Mexico | 10008 | 5.02
| PEMNTVTY | country of birth mother | 199523 | 0 | 43 | United-States | 160479 | 80.43 | Mexico | 9781 | 4.90
| PENATVTY | country of birth self | 199523 | 0 | 43 | United-States | 176989 | 88.71 | Mexico | 5767 | 2.89
| PRCITSHP | citizenship | 199523 | 0 | 5 | Native- Born in the United States | 176992 | 88.71 | Foreign born- Not a citizen of U S  | 13401 | 6.72
| VETQVA | fill inc questionnaire for veterans admin | 199523 | 0 | 3 | Not in universe | 197539 | 99.01 | No | 1593 | 0.80
| PTOTVALB | PTOTVALB | 199523 | 0 | 2 | - 50000. | 187141 | 93.79 | 50000+. | 12382 | 6.21


In [9]:
#(df['AAGE']==32).sum()#32,73
modes= df['AAGE'].value_counts().keys()
card = df['AAGE'].size
mode1= modes[0]
mode2= modes[1]
mode1count= df['AAGE'].value_counts()[mode1]
mode2count= df['AAGE'].value_counts()[mode2]
mode1percent = "{:.2f}".format(mode1count*100.0/float(card))
mode2percent = "{:.2f}".format(mode2count*100.0/float(card))
print(mode1,mode1count,mode1percent,mode2,mode2count,mode2percent)

34 3489 1.75 35 3450 1.73


In [10]:
series = df['AAGE']
card = df['AAGE'].size
min = '{}'.format(series.min())
q1 = '{}'.format(series.quantile(0.25))
median = '{}'.format(series.median())
q3 = '{}'.format(series.quantile(0.75))
max = '{}'.format(series.max())
mean = '{:.2f}'.format(series.mean())
stdev = '{:.2f}'.format(series.std()) 
#var = '{:.4f}'.format(series.var())

#| Feature | Desc. | Count | % of Missing | Card. | Min. | Q1 | Median | Q3 | Max. | Mean | Std. Dev. | Notes |
print(card,"|",
      min,"|",
      q1,"|",
      median, "|",
      q3,"|",
      max,"|",
      mean,"|",
      stdev,"|"
      )

199523 | 0 | 15.0 | 33.0 | 50.0 | 90 | 34.49 | 22.31 |


### Data Quality Report for Cont Variables

| Feature | Desc. | Count | % of Missing | Card. | Min. | Q1 | Median | Q3 | Max. | Mean | Std. Dev. | Notes |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| AAGE | Age | 199,523 | 0 | 91 | 0 | 15 | 33 | 50 | 90 | 34.49 | 22.31 |  |
| AAGE | Age | 199523 | 0 | 91 | 0 | 15.0 | 33.0 | 50.0 | 90 | 34.49 | 22.31 |
| ADTIND | industry code | 199523 | 0 | 52 | 0 | 0.0 | 0.0 | 33.0 | 51 | 15.35 | 18.07 |
| ADTOCC | occupation code | 199523 | 0 | 47 | 0 | 0.0 | 0.0 | 26.0 | 46 | 11.31 | 14.45 |
| AHRSPAY | wage per hour | 199523 | 0 | 1240 | 0 | 0.0 | 0.0 | 0.0 | 9999 | 55.43 | 274.90 |
| CAPGAIN | capital gains | 199523 | 0 | 132 | 0 | 0.0 | 0.0 | 0.0 | 99999 | 434.72 | 4697.53 |
| CAPLOSS | capital losses | 199523 | 0 | 113 | 0 | 0.0 | 0.0 | 0.0 | 4608 | 37.31 | 271.90 |
| DIVVAL | divdends from stocks | 199523 | 0 | 1478 | 0 | 0.0 | 0.0 | 0.0 | 99999 | 197.53 | 1984.16 |
| MARSUPWT | instance weight | 199523 | 0 | 99800 | 37.87 | 1061.615 | 1618.31 | 2188.61 | 18656.3 | 1740.38 | 993.77 |
| NOEMP | num persons worked for employer | 199523 | 0 | 7 | 0 | 0.0 | 1.0 | 4.0 | 6 | 1.96 | 2.37 |
| SEOTR | own business or self employed | 199523 | 0 | 3 | 0 | 0.0 | 0.0 | 0.0 | 2 | 0.18 | 0.55 |
| VETYN | veterans benefits | 199523 | 0 | 3 | 0 | 2.0 | 2.0 | 2.0 | 2 | 1.51 | 0.85 |
| WKSWORK | weeks worked in year | 199523 | 0 | 53 | 0 | 0.0 | 8.0 | 52.0 | 52 | 23.17 | 24.41 |
| YEAR | YEAR | 199523 | 0 | 2 | 94 | 94.0 | 94.0 | 95.0 | 95 | 94.50 | 0.50 |

1.   List item

*   List item
*   List item


2.   List item



### Data Quality Report for Categorical Variables

| Feature | Desc. | Count | % of Missing | Card. | Mode | Mode Freq. | Mode % | 2nd Mode | 2nd Mode Freq. | 2nd Mode Perc | Notes |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| ACLSWKR | Class of worker | 199,523 | 0 | 9 | Not in Universe | 100,245 | 50.24 | Private | 72,028 |  36.10 |  | *italicized text*
| ACLSWKR | class of worker | 199523 | 0 | 9 | Not in universe | 100245 | 50.24 | Private | 72028 | 36.10
| AHGA | education | 199523 | 0 | 17 | High school graduate | 48407 | 24.26 | Children | 47422 | 23.77
| AHSCOL | enrolled in edu inst last wk | 199523 | 0 | 3 | Not in universe | 186943 | 93.69 | High school | 6892 | 3.45
| AMARITL | marital status | 199523 | 0 | 7 | Never married | 86485 | 43.35 | Married-civilian spouse present | 84222 | 42.21
| AMJIND | major industry code | 199523 | 0 | 24 | Not in universe or children | 100684 | 50.46 | Retail trade | 17070 | 8.56
| AMJOCC | major occupation code | 199523 | 0 | 15 | Not in universe | 100684 | 50.46 | Adm support including clerical | 14837 | 7.44
| ARACE | mace | 199523 | 0 | 5 | White | 167365 | 83.88 | Black | 20415 | 10.23
| AREORGN | hispanic Origin | 199523 | 874 | 10 | All other | 171907 | 86.16 | Mexican-American | 8079 | 4.05
| ASEX | sex | 199523 | 0 | 2 | Female | 103984 | 52.12 | Male | 95539 | 47.88
| AUNMEM | member of a labor union | 199523 | 0 | 3 | Not in universe | 180459 | 90.45 | No | 16034 | 8.04
| AUNTYPE | reason for unemployment | 199523 | 0 | 6 | Not in universe | 193453 | 96.96 | Other job loser | 2038 | 1.02
| AWKSTAT | full or part time employment stat | 199523 | 0 | 8 | Children or Armed Forces | 123769 | 62.03 | Full-time schedules | 40736 | 20.42
| FILESTAT | tax filer status | 199523 | 0 | 6 | Nonfiler | 75094 | 37.64 | Joint both under 65 | 67383 | 33.77
| GRINREG | region of previous residence | 199523 | 0 | 6 | Not in universe | 183750 | 92.09 | South | 4889 | 2.45
| GRINST | state of previous residence | 199523 | 0 | 51 | Not in universe | 183750 | 92.09 | California | 1714 | 0.86
| HHDFMX | detailed household and family stat | 199523 | 0 | 38 | Householder | 53248 | 26.69 | Child <18 never marr not in subfamily | 50326 | 25.22
| HHDREL | detailed household summary in household | 199523 | 0 | 8 | Householder | 75475 | 37.83 | Child under 18 never married | 50426 | 25.27
| MIGMTR1 | migration code-change in msa | 199523 | 0 | 10 | ? | 99696 | 49.97 | Nonmover | 82538 | 41.37
| MIGMTR3 | migration code-change in reg | 199523 | 0 | 9 | ? | 99696 | 49.97 | Nonmover | 82538 | 41.37
| MIGMTR4 | migration code-move within reg | 199523 | 0 | 10 | ? | 99696 | 49.97 | Nonmover | 82538 | 41.37
| MIGSAME | live in this house 1 year ago | 199523 | 0 | 3 | Not in universe under 1 year old | 101212 | 50.73 | Yes | 82538 | 41.37
| MIGSUN | migration prev res in sunbelt | 199523 | 0 | 4 | ? | 99696 | 49.97 | Not in universe | 84054 | 42.13
| PARENT | family members under 18 | 199523 | 0 | 5 | Not in universe | 144232 | 72.29 | Both parents present | 38983 | 19.54
| PEFNTVTY | country of birth father | 199523 | 0 | 43 | United-States | 159163 | 79.77 | Mexico | 10008 | 5.02
| PEMNTVTY | country of birth mother | 199523 | 0 | 43 | United-States | 160479 | 80.43 | Mexico | 9781 | 4.90
| PENATVTY | country of birth self | 199523 | 0 | 43 | United-States | 176989 | 88.71 | Mexico | 5767 | 2.89
| PRCITSHP | citizenship | 199523 | 0 | 5 | Native- Born in the United States | 176992 | 88.71 | Foreign born- Not a citizen of U S  | 13401 | 6.72
| VETQVA | fill inc questionnaire for veterans admin | 199523 | 0 | 3 | Not in universe | 197539 | 99.01 | No | 1593 | 0.80
| PTOTVALB | PTOTVALB | 199523 | 0 | 2 | - 50000. | 187141 | 93.79 | 50000+. | 12382 | 6.21


### Q2 - Identify the data quality issues for each attribute. Create a data quality plan and implement the changes. [35 pts] 

Identify missing values, potential anomalies and outliers. Finally, remedy these issues.

In [11]:
# your code goes here

*italicized text*### Q3 - For all the interval and ratio scale features, create a scatter plot matrix (pairplots) and correlation matrix. For all ordinal scale features, create a correlation matrix using Spearman correlation. [35 pts]

Use the data scales list you created in Q1.

In [12]:
# your code goes here